### ISSUE TITLE
Tiny Turtle (who also believes in you)

### ISSUE THREAD

**Issue by plt-amy** [SquidDev-CC/ccjam-2016#80]
![](https://abby.amulet.works/img/64bfa.png)
![](https://abby.amulet.works/img/992bf.png)

**Comment by SquidDev** [comment]
Here's my proposal. It's silly, but the whole feature is:

A tiny turtle is a less powerful, but more cute (and cuddly) version of a normal turtle. The crafting recipe will reflect that: consuming wool or string instead of iron.

The primary purpose of tiny turtles is to look adorable. We will need to have some discussion about how best to achieve this, but I have a couple of proposals:

 - Tiny: think tiny chests, but turtles.
 - Particle effects: there should be a random chance of tiny turtles emitting hearts each tick.
 - Oversized tools: I'm not entirely sure if this'll work, but if we can make it look like the tiny turtle is using tools too large for it, I think it'll be kinda cute.
 - Googly eyes: or some other optional accessories. Basically I just want to put [this][boppy] on a turtle.
 - More?

I realise there's a risk of just copying CCEdu features here. I'm not entirely adverse to that, but I don't want to just become a rip-off.

For those people with no joy in their life, tiny turtles also have several unique attributes which may make them more suited to particular tasks. Though that's basically child labour, and you wouldn't do that right?

 - Tiny turtles move and turn faster. We will need to do some work to find a speed which is still useful but not jerky.
 - **[Not sure]** Tiny turtles dig slightly slower.
 - More?

[boppy]: https://cdn.shopify.com/s/files/1/1555/2777/products/gemstone_boppy_purple_headband_image_2_1024x1024.jpg?v=1478807562

**Comment by Lignum** [comment]
> Tiny turtles move and turn faster. We will need to do some work to find a speed which is still useful but not jerky.
[Not sure] Tiny turtles dig slightly slower.

A common problem we have with ComputerCraft is that turtles are too slow for things like mining for them to be effective. So, how about we make them dig faster as well, in addition to moving and turning? 

In exchange, they'd have to be more expensive, of course. Something about packing more stuff in a smaller space, like with modern processors?

**Comment by Lemmmy** [comment]
if you have less body mass to move, you can move it quicker...or something?

**Comment by Lignum** [comment]
Of course, but dig quicker?

**Comment by Lemmmy** [comment]
hrrrm, that'd require more force, and usually you'd depend on your own mass versus gravity to get more force i think
so probably not?

**Comment by SquidDev** [comment]
@Lemmmy Tiny children are more excitable and hyperactive. Tiny turtles are just the same!

**Comment by plt-amy** [comment]
I don't think the cute turtles should have any use at all, other than maybe `dance`. Maybe make them be able to do stuff, but make an "ouch" sound (the player hurt effect) every time they dig?

**Comment by plt-amy** [comment]
So you feel bad and don't make them do stuff, that is.

**Comment by Lemmmy** [comment]
i fucking hate tiny children :)

**Comment by Lemmmy** [comment]
can we make them killable

**Comment by ghost** [comment]
just make it swing faster but take more hits to mine something

**Comment by Lemmmy** [comment]
who even are you

**Comment by ybden** [comment]
idk, but you're probably on some list now

**Comment by plt-amy** [comment]
@ybden Turtles always break blocks instantly, though.

**Comment by Lemmmy** [comment]
ive been on several lists for quite a few years

**Comment by ybden** [comment]
ah well, always an opportunity for some more

**Comment by SquidDev** [comment]
It's possible to add more incremental block breaking (see turtle tool host, kinetic augment, etc...) but it's a bit of a pain. 

Whilst I'd love to have more useful (read overpowered) turtles, I'd kinda like to keep tiny turtles to a silly feature. Anyway, given you can put lasers on them I don't think block break speed is exactly low. 


Though with regards to turtle breaking speed, I have had some ideas for a magic/CC crossover: making turtles respond to beacons, potions, etc... It's gonna require some substantial hooks to CC first though.

**Comment by Lemmmy** [comment]
and cause 1500 bugs with sponge :)

**Comment by ghost** [comment]
@SquidDev regarding instabreak, would it be particularly difficult to make tiny turtles take multiple hits to break something?

**Comment by zardyh** [comment]
@Lemmmy that's implied

**Comment by plt-amy** [comment]
Could we just have the tiny turtles be full-on entities instead of the pseudo-blocks they are now?

**Comment by plt-amy** [comment]
HEY! THE POTATO PART IS IMPORTANT, SQUID. _I NEED A TURTLE THAT BELIEVES IN ME._

**Comment by ybden** [comment]
$$$$$$ here's some feeback, mister

**Comment by Lemmmy** [comment]
good idea: make the tiny turtles only support urn

**Comment by zardyh** [comment]
i approve

**Comment by SquidDev** [comment]
Wait, you can undo my edits? Hrrmr. 

> Could we just have the tiny turtles be full-on entities instead of the pseudo-blocks they are now?

Making the turtle an entity would be possible, but I try to avoid going near entity code if I can. I'd kinda like to keep to the turtle aesthetic if possible, and that probably means sticking with a block. That doesn't mean I have to keep any of the turtle code though, so it should be possible to get things to run more smoothly.

> regarding instabreak, would it be particularly difficult to make tiny turtles take multiple hits to break something?

Most of the code is already in Plethora TBH (though for entities rather than turtles) and so it would be possible. I do kinda like the mental image of a turtle swinging super fast but rather ineffectually at a block.

**Comment by zardyh** [comment]
> Wait, you can undo my edits? Hrrmr.

No, but I can also edit.

**Comment by CrazedProgrammer** [comment]
🍆👍

**Comment by SquidDev** [comment]
Yes, I'm an awful person. But it makes it easier to separate useful discussion from casual chat. Sorry :(.

**Comment by Lemmmy** [comment]
its ok you are great

**Comment by BTCTaras** [comment]
What if they had a smaller 3x3 inventory like turtles did a long time ago? Possibly opening up compatibility with ancient turtle scripts that involve crafting.

**Comment by SquidDev** [comment]
> What if they had a smaller 3x3 inventory like turtles did a long time ago

It's something I've thought about, but I think too many programs hard-code the inventory size as 16 for this to work well.

**Comment by Lemmmy** [comment]
hey - all the more reason to add turtle.getInventorySize()

**Comment by hugeblank** [comment]
Here's a little bit of a derpy proposal, what if the tiny turtle wasn't properly used in the physical world except for maybe a decoration. 

Just a couple of ideas that could be features:
- The tiny turtle would be used in any storage inventory (chests, player invo, turtles, etc.)
- It is able to shuffle up down left and right in the inventory
- It has a little 3x3 inventory, and functions that allow it to suck from any direction
- It can have all the peripheral tools equipped to itself
  - it can break blocks that are in your inventory into their respective materials using a pickaxe (diamond ore to diamonds)
  - it can attack with a sword... if it's in a player inventory this would hit the player (but with 0 damage), and otherwise have no effect.
- It can put itself into permanent inventories (shulker box, ender chest) to hide. Being tiny is scary. (Note: this would probably be a one time thing, I don't think that it would be very easy to 'unhide' it and get it out of the permanent inventory. Plus it would add to the whole being cute and tiny feel, it refuses to come back out once its hidden.)

This idea has a few advantages:
1. It serves a functional purpose
2. It can also be decorational when placed into the world
3. It's the cutest darn in inventory based 3x3 crafting table ever

**Comment by SquidDev** [comment]
@hugeblank Whilst it's a cool idea, it's nigh-on impossible to implement within the constraints of Minecraft's codebase:

 - Firstly, items inside an inventory aren't ticked. Whilst it is _technically_ possible to get around that by looping through every tile entity and then every item, that's going to get very laggy very quickly.
 - Secondly, the inventory itself doesn't really have any concept of "layout" - it's just a list of slots. The actual positions are determined by the 'container', which is a massive pain to get as you'd have to simulate a right click, then hook into whatever container your fake player opened.

It'd definitely be possible to just do this inside player inventories, as they have a known size, and are ticked, but that's a little more useless - you can just use the neural interface to achieve the same tasks.

**Comment by Jummit** [comment]
Imagine running the `excavate` program on a tiny turtle in your inventory to clear up space.

**Comment by Lemmmy** [comment]
ok

**Comment by BTCTaras** [comment]
This recent forum suggestion might be interesting: http://www.computercraft.info/forums2/index.php?/topic/29529-monitor-on-turtles/

Tiny turtles with a tiny screen on them. It would give them use cases above not-as-tiny turtles

**Comment by SquidDev** [comment]
> Monitor on Turtles

Yeah, I saw that and did think about it. It's something which would be insanely cool, but I'd want to have a thunk about implementation. 

As a turtle moving pretty much boils down to "break and replace the TE", you end up sending an awful lot of information to the client each move. Adding a terminal (even turtle's small ones) would bulk out the packet a bit (just under 1kb for a 13x39 screen).

Truth be told, it probably isn't much of an issue, as monitors redrawing a lot have a much worse impact on network usage, but it's worth bearing in mind.

**Comment by shelbykauth** [comment]
Hey, for the tiny turtle who loves you, maybe they don't have any inventory.  Or just one slot.  And they are primarily for moving one item at a time or moving around and turning on redstone.

**Comment by shelbykauth** [comment]
Oooooh!  I just had a thought about the turtle screen thing.  Instead of keeping text on the screen when the turtle moves, the text disappears every time like it does with monitors. So it only draws once unless you print more stuff out.

**Comment by Jummit** [comment]
[Cat ears?](https://pm1.narvii.com/5836/debacb6f7b4e3a0ca2da14936036646566b9a62a_hq.jpg)

**Comment by ghost** [comment]
tfw still getting notifications on this
